### DISCUSSION TITLE
Color scheme with more than 10 colors?

### DISCUSSION THREAD

**Discussion by knutwannheden** [observablehq/plot#1072] [Q&A] [answered]
Like in the example https://observablehq.com/@observablehq/plot-cheatsheets-colors?collection=@observablehq/plot-cheatsheets#colorsExample I have categorical data with more than 10 discrete values. Now I am wondering if there is a color scheme which I can use that would already provide this. Or should I "manually" map the domain to a fixed range of colors?

I understand that the colors may end up being hard to discern, but I think in the exploratory phase it would be very useful nonetheless.

**Comment by Fil** [comment, marked as answer]
You can specify the color range as an array; for example, to reuse the (deprecated) Category20b from d3@4, you can specify:

`color: {range: ["#393b79","#5254a3","#6b6ecf","#9c9ede","#637939","#8ca252","#b5cf6b","#cedb9c","#8c6d31","#bd9e39","#e7ba52","#e7cb94","#843c39","#ad494a","#d6616b","#e7969c","#7b4173","#a55194","#ce6dbd","#de9ed6"] }
`

**Reply by knutwannheden** [reply]
Thanks. I was hoping that there would be some utility function to "generate" colors. If I could plug in a function I could imagine a function returning values by always moving some fixed step length along the V dimension in HSV space.

**Reply by mbostock** [reply]
You can use one of the ordinal schemes to create more than 20 colors. But ordinal schemes are ordered, as the name implies, so they typically don’t make very good categorical schemes.

![untitled (62)](https://user-images.githubusercontent.com/230541/194801957-9647f206-af82-4a7c-90fe-78ba0e38ad24.png)

```js
Plot.plot({
  color: {
    scheme: "spectral"
  },
  marks: [
    Plot.cellX("ABCDEFGHIJKLMNOPQSTUVWXYZ")
  ]
})
```

If you specify the color scale’s range as a function, it will be passed the color domain, and so you can also write a custom interpolator to create a color scheme dynamically. The above is equivalent to:

![untitled (63)](https://user-images.githubusercontent.com/230541/194802029-5ca27f17-6d3b-4d7b-ae63-cbcc03d7c3df.png)

```js
Plot.plot({
  color: {
    range: ({length}) => d3.quantize(d3.interpolateSpectral, length)
  },
  marks: [
    Plot.cellX("ABCDEFGHIJKLMNOPQSTUVWXYZ")
  ]
})
```

You could replace d3.interpolateSpectral with something that generates an HSV color, for example. (D3 doesn’t have built-in support for HSV, but you can use another library such as d3-hsv.)

**Reply by knutwannheden** [reply]
Excellent!
